### PR TITLE
test(zoe): Add a test for contracts surviving Zoe vat upgrade

### LINKED ISSUE #6713: Missing tests for contracts surviving Zoe vat upgrade
## What is the Problem Being Solved?

#5997 made Zoe durable, but we still need to test that upgrading it doesn't break active contract instances.

## Test Plan

Create a test that instantiates a contract, upgrades Zoe, and then interacts with the contract.

### PR DESCRIPTION
**Pull request by gibson042**
closes: #6713

## Description

Creates a test that instantiates covered-call contracts, null-upgrades Zoe, and then finishes the contracts.

### Security Considerations

n/a

### Documentation Considerations

n/a

### CODE REVIEW

File: packages/zoe/test/swingsetTests/zoe/test-zoe-upgrade.js [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+  // continuation after upgrade.
+  const pausedFlows = [];
+  for (let i = 0; i < flow.length; i += 1) {
+    const [beforeStepName] = flow[i];
+    const result = await doSteps(`pre-${beforeStepName}`, flow.slice(0, i));
+    pausedFlows.push({ result, remainingSteps: flow.slice(i) });
+  }
+
+  // Null-upgrade Zoe.
+  const incarnationNumber = await run('upgradeVat', [zoeVatConfig]);
```
Comment (gibson042) [inline packages/zoe/test/swingsetTests/zoe/test-zoe-upgrade.js]: @Chris-Hibbert I think I will need some help after all.

**Error: vat-upgrade failure**
```
error during vat dispatch() of startVat,[object Object] (Error#1)
Error#1: defineDurableKind not called for tags: [ZDEFAULT brand,ZDEFAULT payment,ZDEFAULT Purse,ZDEFAULT issuer,ZDEFAULT mint,FeeMint,Zoe Invitation brand,Zoe Invitation payment,Zoe Invitation Purse,Zoe Invitation issuer,Zoe Invitation mint,InstanceAdmin,BundleIDInstallation,BundleInstallation,InstallationStorage,ZoeMint,ZoeStorageManager,InstanceHandle,zoe Seat publisher,ZoeSeatKit,SeatHandle,instanceAdmin,zoeInstanceAdmin,emptyCreatorFacet,emptyPublicFacet,adminFacet,ZoeService,InstanceStorageManager,InstanceStorageManager,InstanceStorageManager,InstanceStorageManager,InstanceStorageManager,InstanceStorageManager]
Error: defineDurableKind not called for tags: [ZDEFAULT brand,ZDEFAULT payment,ZDEFAULT Purse,ZDEFAULT issuer,ZDEFAULT mint,FeeMint,Zoe Invitation brand,Zoe Invitation payment,Zoe Invitation Purse,Zoe Invitation issuer,Zoe Invitation mint,InstanceAdmin,BundleIDInstallation,BundleInstallation,InstallationStorage,ZoeMint,ZoeStorageManager,InstanceHandle,zoe Seat publisher,ZoeSeatKit,SeatHandle,instanceAdmin,zoeInstanceAdmin,emptyCreatorFacet,emptyPublicFacet,adminFacet,ZoeService,InstanceStorageManager,InstanceStorageManager,InstanceStorageManager,InstanceStorageManager,InstanceStorageManager,InstanceStorageManager]
 at apply ()
 at Error (/bundled-source/.../node_modules/ses/src/error/tame-error-constructor.js:54)
 at insistAllDurableKindsReconnected (/bundled-source/.../packages/SwingSet/src/liveslots/virtualObjectManager.js:994)
 at startVat (/bundled-source/.../packages/SwingSet/src/liveslots/liveslots.js:1437)
 at ()
 at ()
```

File: packages/zoe/test/swingsetTests/upgradeCoveredCall/vat-ertp-service.js [hunk 2]
```diff
@@ -1,9 +1,12 @@
 import { Far } from '@endo/marshal';
+import { AmountMath } from '@agoric/ertp';
 import { vivifyErtpService } from '@agoric/ertp/test/swingsetTests/ertpService/vat-ertp-service.js';
 
 export const buildRootObject = async (vatPowers, _vatParams, baggage) => {
+  const exportableAmountMath = Far('AmountMath', { ...AmountMath });
```
Comment (Chris-Hibbert) [inline packages/zoe/test/swingsetTests/upgradeCoveredCall/vat-ertp-service.js:6]: Why do we need an `exportableAmountMath`?  AmountMath was designed to be usable locally in every vat and produce copy objects that can be sent around seamlessly.

File: packages/zoe/test/swingsetTests/zoe/test-zoe-upgrade.js [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+      return kunser(result);
+    }
+    assert(status === 'rejected');
+    const err = c.kpResolution(kpid);
+    throw kunser(err);
+  };
+  const messageVat = (name, methodName, args) =>
+    run('messageVat', [{ name, methodName, args }]);
+  const messageObject = (presence, methodName, args) =>
+    run('messageVatObject', [{ presence, methodName, args }]);
```
Comment (Chris-Hibbert) [inline packages/zoe/test/swingsetTests/zoe/test-zoe-upgrade.js:55]: Are these a replacement for `E()` calls?  What role do they play?

**Review by Chris-Hibbert — COMMENTED** [review]
Thanks, @gibson042, I'll take it from here for now. This is a great framework for verifying upgradeability, and I'll use it to figure out what I didn't set up correctly for upgrade.

**Comment by gibson042** [comment]
> Are these a replacement for `E()` calls? What role do they play?

> Why do we need an `exportableAmountMath`? AmountMath was designed to be usable locally in every vat and produce copy objects that can be sent around seamlessly.

The answer to both questions is that dependent code exists _outside_ of the vats under test and is thus limited to pass-by-copy data as supported by encoding and decoding in packages/SwingSet/test/bootstrap-relay.js (which does use `E` directly).

**Comment by Chris-Hibbert** [comment]
@gibson042, I pushed one teensy correction. Otherwise, the test worked great!  #6726 contains the changes to make Zoe pass the test. This test is now stacked on top of that.

**Comment by erights** [comment]
Are these two CI dapp failures due to the API change not yet being propagated to other repos? If so, why didn't they already fail in #6726 ?

**Comment by Chris-Hibbert** [comment]
I hadn't bothered to add the markup below to this one, since it's a known fix. Once the base PR is merged, this will get it.
```
#dapp-otc-branch: 5997-integration
#documentation-branch: 5997-integration
```
I'll add it above, and if the tests run from scratch, they'll get it.

**Comment by gibson042** [comment]
The `incarnationNumber` change should not be necessary given the current state of [`upgradeVat` in master](https://github.com/Agoric/agoric-sdk/blob/19855de77cca7a96c1c3a42f090502d11ba36989/packages/SwingSet/test/bootstrap-relay.js#L71-L78), but I'm happy to let that shake out as appropriate. I'll review #6726 today.

**Comment by Chris-Hibbert** [comment]
I remove the cross-repo imports in the top comment. I'll rebase this on the latest #6726 to see that the tests still pass.

**Comment by Chris-Hibbert** [comment]
The code this is a test for has been merged. @gibson042, please merge this.

**Review by Chris-Hibbert — APPROVED** [review]
LGTM

**Pull request merged**